packages/backend/src/services/ServiceRepository.ts

```diff
@@ -94,111 +159,222 @@ abstract class ServiceRepositoryBase {
+            'pinningService',
+            () => this._services.pinningService,
+        );
     }
 
     public getProjectService(): ProjectService {
-        return this._services.projectService;
+        return this.getService(
+            'projectService',
+            () => this._services.projectService,
+        );
     }
 
     public getSavedChartService(): SavedChartService {
-        return this._services.savedChartService;
+        return this.getService(
+            'savedChartService',
+            () => this._services.savedChartService,
+        );
     }
 
     public getSchedulerService(): SchedulerService {
-        return this._services.schedulerService;
+        return this.getService(
+            'schedulerService',
+            () => this._services.schedulerService,
+        );
     }
 
     public getSearchService(): SearchService {
-        return this._services.searchService;
+        return this.getService(
+            'searchService',
+            () => this._services.searchService,
+        );
     }
 
     public getShareService(): ShareService {
-        return this._services.shareService;
+        return this.getService(
+            'shareService',
+            () => this._services.shareService,
+        );
     }
 
     public getSshKeyPairService(): SshKeyPairService {
-        return this._services.sshKeyPairService;
+        return this.getService(
+            'sshKeyPairService',
+            () => this._services.sshKeyPairService,
+        );
     }
 
     public getSpaceService(): SpaceService {
-        return this._services.spaceService;
+        return this.getService(
+            'spaceService',
+            () => this._services.spaceService,
+        );
     }
 
     public getUnfurlService(): UnfurlService {
-        return this._services.unfurlService;
+        return this.getService(
+            'unfurlService',
+            () => this._services.unfurlService,
+        );
     }
 
     public getUserAttributesService(): UserAttributesService {
-        return this._services.userAttributesService;
+        return this.getService(
+            'userAttributesService',
+            () => this._services.userAttributesService,
+        );
     }
 
     public getUserService(): UserService {
-        return this._services.userService;
+        return this.getService('userService', () => this._services.userService);
     }
 
     public getValidationService(): ValidationService {
-        return this._services.validationService;
+        return this.getService(
+            'validationService',
+            () => this._services.validationService,
+        );
+    }
+
+    /**
+     * Handles initializing a service, and taking into account service
+     * providers + memoization.
+     *
+     * If a factory is not provided, and a service provider is not defined,
+     * this method throws an error. This should not happen in normal operation.
+     */
+    public getService<
```

Comment:
Minor: should this be private or is there a use case for this method to be used outside?

Reply:
I would even consider moving this and the `_serviceCache` to `ServiceRepositoryBase` 🤔 
This way this class is only for the service methods.

Reply:
Good catch, shouldn't be public!

Reply:
Re: moving it to `ServiceRepositoryBase`, I agree on principle - but the method needs to be aware of the actual service definitions, since we want to pass `ServiceRepository` down to the providers once it's "finalized" (so that providers can also depend on services from the repository).

Leaving the method as part of `ServiceRepository` is the most straight-forward way to do this.